poly/promise/promise.go

```diff
@@ -0,0 +1,1355 @@
+	s += "t => " + strconv.Itoa(p.t) + ",\n"
+	s += "r => " + strconv.Itoa(p.r) + ",\n"
+	s += "n => " + strconv.Itoa(p.n) + ",\n"
+	s += "Public Key => " + p.pubKey.String() + ",\n"
+	s += "Public Polynomial => " + p.pubPoly.String() + ",\n"
+	insurers := ""
+	secrets := ""
+	for i := 0; i < p.n; i++ {
+		insurers += p.insurers[i].String() + ",\n"
+		secrets += p.secrets[i].String() + ",\n"
+	}
+	s += "Insurers =>\n[" + insurers + "],\n"
+	s += "Secrets =>\n[" + secrets + "]\n"
+	s += "}\n"
+	return s
+}
+
+/* The PromiseState struct is responsible for maintaining state about Promise
+ * structs. It consists of three main pieces:
+ *
+ *    1. The promise itself, which should be treated like an immutable object
+ *    2. The shared secrets the server has recovered so far
+ *    3. A list of signatures from insurers cerifying the promise
+ *
+ * Each server should have one PromiseState per Promise
+ *
+ * Note to users of this code:
+ *
+ *    To add a share to PriShares, do:
+ *
+ *       p.PriShares.SetShare(index, share)
+ *
+ *    To reconstruct the secret, do:
+ *
+ *       p.PriShares.Secret()
+ *
+ *    Be warned that Secret will panic unless there are enough shares to
+ *    reconstruct the secret. (See poly/sharing.go for more info)
+ *
+ * TODO Consider if it is worth adding a String function
+ */
+type PromiseState struct {
+
+	// The actual promise
+	Promise Promise
+
+	// Primarily used by clients, contains shares the client has currently
+	// obtained from insurers. This is what will be used to reconstruct the
+	// promised secret.
+	PriShares poly.PriShares
+
+	// A list of signatures verifying that an insurer has cerified the
+	// secret share it is guarding.
+	signatures []*PromiseSignature
+
+	// A list of blame proofs in which an insurer proves its secret share
+	// to be malformed and hence proves the promiser of being malicious
+	blames []*BlameProof
+}
+
+/* Initializes a new PromiseState
+ *
+ * Arguments
+ *    promise = the promise to keep track of
+ *
+ * Returns
+ *   An initialized PromiseState
+ */
+func (ps *PromiseState) Init(promise Promise) *PromiseState {
+	ps.Promise = promise
+
+	// Initialize a new PriShares based on information from the promise.
+	ps.PriShares = poly.PriShares{}
+	ps.PriShares.Empty(promise.shareSuite, promise.t, promise.n)
+
+	// There will be at most n signatures and blame proofs, one per insurer
+	ps.signatures = make([]*PromiseSignature, promise.n, promise.n)
+	ps.blames = make([]*BlameProof, promise.n, promise.n)
+	return ps
+}
+
+/* Adds a signature from an insurer to the PromiseState
+ *
+ * Arguments
+ *    i   = the index in the signature array this signature belongs
+ *    sig = the PromiseSignature to add
+ *
+ * Postcondition
+ *   The signature has been added
+ *
+ * Note to users of this code
+ *   Be sure to call ps.Promise.VerifySignature before calling this function
+ */
+func (ps *PromiseState) AddSignature(i int, sig *PromiseSignature) {
```

Comment:
As per above, there should only be one 'AddSignature' or 'AddResponse' or some such, for a single public type that can represent either a signature (attesting that the promise was good) or a blame-proof (attesting that the promise was bad).